src/io/OhmmsData/XMLParsingString.h

```diff
@@ -50,41 +89,60 @@ class XMLNodeString : public std::string
   }
 };
 
-/** convert an xmlNode attribute into a std::string
+/** convert an xmlNode attribute into a name value pair.
+ * used to inherit from std::string but to no advantage
+ * and repeated confusion about whether comparisons to an attribute string were between
+ * name or value etc.
  * if parsing multiple attributes is needed, use OhmmsAttributeSet
  */
-class XMLAttrString : public std::string
+class XMLAttrString
 {
+  // ordering same as crazy value, name signature of constructor
+  std::string attribute_;
   const std::string attribute_name_;
-
 public:
   /// construct a string from an xmlNode
   XMLAttrString(const xmlNodePtr cur, const char* name)
       : attribute_name_(name)
   {
-    xmlChar* attr_char = xmlGetProp(cur, (const xmlChar*)name);
+    xmlChar* attr_char = xmlGetProp(cur, castCharToXMLChar(name));
     if(attr_char)
     {
-      assign((const char*)attr_char);
+      attribute_.assign(castXMLCharToChar(attr_char));
       xmlFree(attr_char);
     }
   }
-
+  
   /// expose base class constructors
-  XMLAttrString(const std::string& in, const std::string& name) : std::string(in), attribute_name_(name) { }
+  XMLAttrString(const std::string& in, const std::string& name) : attribute_(in), attribute_name_(name) { }
 
-  XMLAttrString(const char* in, const char* name) : std::string(in), attribute_name_(name) { }
+  XMLAttrString(const char* in, const char* name) : attribute_(in), attribute_name_(name) { }
 
+  /** i.e. does it have a value
+   */
+  auto hasValue() const { return ! attribute_.empty(); };
+  
   /// write a string to an xmlNode
   void createXMLAttribute(xmlNodePtr cur) const
   {
-    xmlNewProp(cur, (const xmlChar*)attribute_name_.c_str(), (const xmlChar*)this->c_str());
+    xmlNewProp(cur, castCharToXMLChar(attribute_name_.c_str()), castCharToXMLChar(attribute_.c_str()));
   }
 
   /// write a string to an xmlNode
   void setXMLAttribute(xmlNodePtr cur) const
   {
-    xmlSetProp(cur, (const xmlChar*)attribute_name_.c_str(), (const xmlChar*)this->c_str());
+    xmlSetProp(cur, castCharToXMLChar(attribute_name_.c_str()), castCharToXMLChar(attribute_.c_str()));
   }
+
+  /** retain surprising behavior where this inherently name value object 
+   *  can also operate as just a string of its value
+   *  \todo remove this its a footgun and makes for less readable code
+   */
+  operator const std::string&() { return attribute_; }
```

Comment:
Now. I'm confused. You turned the base class to a member because of confusion. Then you provide this convenience conversion which retain the confusion.

I kind of still like the old way. To reduce the confusion. We can call it XMLAttrValueString. Having String in the type name is to say it is derived from std::string and can be used as an std::string object.

With the current modification, the class need to be removed probably as XMLStringAttr indicating it is an XML attribute for string data.

Reply:
I'll just make the extra changes and delete the conversion. 

It's a pair of string buffers. a name value pair. It's methods should make this clear. It could just be XMLAttr. I didn't like the way it interacted with lowerCase so I think its time to change it. Also the fact that it has an attribute_name is often called aname and is most frequently used for attributes with the tag "name" makes me think less implicit here is going to be much more clear.

Subclassing standard stl containers and giving them weird semantics is not good practice. It results in surprising behavior. There is no good reason to subclass this except to get a bunch of cute operators that only effect the value half of the object. Or apply mutating functions for std::string that perhaps surprisingly only effect the value part of the pair.

Reply:
I thought about XMLAttr but it doesn't say the value type is a string or int or float. For this reason, XMLStringAttr comes out.

Reply:
It seems easy to just remove the type altogether I could find anywhere the state is actually used.

Reply:
> It seems easy to just remove the type altogether I could find anywhere the state is actually used.

Confused. Which type? Which state?

Reply:
XMLAttrString which differs from a normal string in that it remembers the name of the attribute it was constructed from.